### ISSUE TITLE
Unable to resolve hostnames from within containers.

### ISSUE THREAD

**Issue by western-hoolock** [klutchell/unbound-docker#204]
I have this image running on a machine (192.168.0.100) in my local network and configured my router's DHCP server to hand out 192.168.0.100 as being the DNS server.
Everything works fine, all my machines (including the host running docker) are able to perform lookups EXCEPT the apps running in other docker containers! I can only do lookups for other container names e.g.: If i have a docker-compose.yml with an app and db service, doing ```ping app```from the db container, and vice versa, works fine. Trying to lookup anything else fails.

If I change my docker-compose.yml for the unbound container from:
```
    ports:
      - 53:53/tcp
      - 53:53/udp
```

to
```
    ports:
      - 53:53/tcp
      - 192.168.0.100:53:53/udp
```

Lookups start working again inside containers. I'm not sure why. It's fine for now but I'd really like to avoid having to hard code the ip address of the docker host in my docker-compose.yml file.

Any insights as to what might be causing this behavior?

**Comment by klutchell** [comment]
Are you by any chance blocking 3rd party DNS at your gateway? Asking because the default DNS for docker containers is usually 8.8.8.8. This can be set [on the daemon](https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-dns-options), or [per container](https://docs.docker.com/compose/compose-file/compose-file-v3/#dns).

**Comment by western-hoolock** [comment]
> Are you by any chance blocking 3rd party DNS at your gateway? Asking because the default DNS for docker containers is usually 8.8.8.8. This can be set [on the daemon](https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-dns-options), or [per container](https://docs.docker.com/compose/compose-file/compose-file-v3/#dns).

No those aren't blocked and I think Docker isn't using them since the docs state:

> Containers that attach to a [custom network](https://docs.docker.com/network/network-tutorial-standalone/#use-user-defined-bridge-networks) use Docker's embedded DNS server. The embedded DNS server forwards external DNS lookups to the DNS servers configured on the host.

The unbound container is attached to it's own network and the DNS server configured on the Docker host (through DHCP) is 192.168.0.100 (i.e. itself).

Some interesting results from within a container attached to the same Docker network unbound is attached to:
(127.0.0.11 is what ```cat /etc/resolv.conf``` shows when done inside the container)
```dig unbound @127.0.0.11``` works.
```dig www.google.com @127.0.0.11``` does not work.
```dig www.google.com @ 192.168.0.100``` does not work either, so probably the internal Docker DNS server is forwarding but just can't reach it?
```dig www.google.com @ 8.8.8.8``` works...

When I say 'does not work' I mean it simply times out.

Everything else on my LAN, including the Docker host can use 192.168.0.100 as the DNS server just fine. I'm stumped as to why binding to just the Docker host's LAN interface makes it work.

**Comment by klutchell** [comment]
Thinking about your two tests...

```
    ports:
      - 53:53/tcp
      - 53:53/udp
```

In this case `0.0.0.0:53:53` is implied and docker will attempt to bind to all interfaces on the host.

```
    ports:
      - 53:53/tcp
      - 192.168.0.100:53:53/udp
```

Unclear why limiting the binding to a single interface like this would allow lookups from other containers.

Are all other containers on the same network as unbound? You said some of them were, but doesn't it make a difference if they aren't? What if they use host networking?

Maybe what's happening is the internal docker network in your case needs to be added to the access control list [here](https://github.com/klutchell/unbound-docker/blob/f226ad91b156534ccfcc46755770dd2e19fab028/rootfs_overlay/etc/unbound/unbound.conf#L98-L103)?

The docker host is forwarding the lookup to unbound but the source IP of the container is not on the control list? I would think that forwarded DNS lookups would not contain the original source IP but worth investigating I guess.

**Comment by western-hoolock** [comment]
> Maybe what's happening is the internal docker network in your case needs to be added to the access control list [here](https://github.com/klutchell/unbound-docker/blob/f226ad91b156534ccfcc46755770dd2e19fab028/rootfs_overlay/etc/unbound/unbound.conf#L98-L103)?

I checked and Docker uses IP address for all the networks/containers that are already in those ranges.

I do have some interesting network sniffer logs though. I started a new container, attached it to the same Docker network the Unbound container is attached to and then do a ```ping www.google.com``` from the new container I just started and log the network traffic.

```
13:36:09.640930 IP 34352cbd3350.35617 > 192.168.0.100.53: 43778+ A? www.google.com. (32)
13:36:09.640930 IP 34352cbd3350.35863 > 192.168.0.100.53: 44968+ AAAA? www.google.com. (32)
13:36:09.667592 IP 192.168.0.100.53 > 34352cbd3350.35617: 43778 1/0/0 A 172.217.168.228 (48)
13:36:09.677338 IP 192.168.0.100.53 > 34352cbd3350.35863: 44968 1/0/0 AAAA 2a00:1450:400e:80d::2004 (60)
```

The log above shows the container (hostname: ```34352cbd3350``` doing A and AAAA lookups for google.com and getting an answer back from 192.168.0.100) everything worked because I explicitly specified the LAN interface when exposing port 53

Now I run the same experiment but not bound to the LAN interface ```53:53/udp``` Note that the Docker network's subnet is ```172.30.0.0/16``` the Unbound container has an ip of ```172.30.0.2``` and the container I'm pinging from has ip ```172.30.0.3``` and hostname ```16435b6890e5```

```
13:31:05.166175 IP 16435b6890e5.53250 > 192.168.0.100.53: 45255+ AAAA? www.google.com. (32)
13:31:05.166179 IP 16435b6890e5.51613 > 192.168.0.100.53: 44291+ A? www.google.com. (32)
13:31:05.166863 IP 172.30.0.1.53 > 16435b6890e5.53250: 45255 1/0/0 AAAA 2a00:1450:400e:80d::2004 (60)
13:31:05.166913 IP 16435b6890e5 > 172.30.0.1: ICMP 16435b6890e5 udp port 53250 unreachable, length 96
13:31:05.167020 IP 172.30.0.1.53 > 16435b6890e5.51613: 44291 1/0/0 A 172.217.168.228 (48)
13:31:05.167065 IP 16435b6890e5 > 172.30.0.1: ICMP 16435b6890e5 udp port 51613 unreachable, length 84
```

Now we see the same A and AAAA requests going to 192.168.0.100 but the Docker network's gateway ```172.30.0.1``` is forwarding the packet and the ping container didn't expect that source address so it sends back an ICMP port unreachable. So the answer just never makes it back.
I then tried a ```dig @172.30.0.2 www.google.com``` and sure enough that worked since the answer came straight from 172.30.0.2 (the Unbound container) and there was no need for it to go through the gateway.

There's a comment to an open issue here https://github.com/moby/moby/issues/15127#issuecomment-323663787 that suggests this is inherent to binding to 0.0.0.0 when using UDP and the app being unable to reliably retrieve the address of the actual interface the packet was received on. So when sending it out it has to pick an interface and apparently it picks the wrong one?
I don't really understand why everything works fine for devices on my LAN when binding to all interfaces though, those replies magically pick the right interface/source address?

**Comment by klutchell** [comment]
> I don't really understand why everything works fine for devices on my LAN when binding to all interfaces though, those replies magically pick the right interface/source address?

I'm a bit out of my depth at this point, but wouldn't that be because LAN devices are not using the moby DNS forwarding and instead speak directly to unbound on the interface?

**Comment by western-hoolock** [comment]
> > I don't really understand why everything works fine for devices on my LAN when binding to all interfaces though, those replies magically pick the right interface/source address?
> 
> I'm a bit out of my depth at this point, but wouldn't that be because LAN devices are not using the moby DNS forwarding and instead speak directly to unbound on the interface?

Yeah I don't know enough about networking/iptables to properly reason about what exactly is going on here. [This blog post](https://blog.powerdns.com/2012/10/08/on-binding-datagram-udp-sockets-to-the-any-addresses) by PowerDNS seems to suggest the server is responsible for picking the right source address and setting it when sending a datagram back.
It looks like Unbound has an (experimental) configuration option named ```interface-automatic``` that should allow it to properly set the source address when binding to all interfaces but that doesn't seem to do anything for my situation.

I have moved on though since this was taking way too much of my time and if you do a bit of Googling a ton of Pihole results come up where people have the same problem and no clear explanation as to what's going on comes up, ever. They just resort to binding to an interface or run Unbound on another machine.
It's definitely not a problem with this image for sure so I'll close this. Thanks for your time!

**Comment by western-hoolock** [comment]
So I ran into this issue again and finally figured out why Docker containers are unable to resolve external hostnames when using this Unbound image:

As mentioned in the earlier posts Docker's DNS server runs on 127.0.0.11 by default and resolves docker container names and forwards everything else to the host's default DNS server. 

However in the unbound.conf file there's this line:
```access-control: 127.0.0.1/32 allow```

Which only allows requests from 127.0.0.1 but the Docker DNS server is at 127.0.0.11 so any requests for outside domains it was forwarding would just get dropped by Unbound.
After adding ```access-control: 127.0.0.11/32``` I was able to use the Unbound server from within other containers no problem without having to explicitly specify the interface to bind to in the docker-compose.yml file.

Perhaps change ```access-control: 127.0.0.1/32 allow``` to ```access-control: 127.0.0.0/24 allow``` or even ```access-control: 127.0.0.0/8 allow```?

**Comment by klutchell** [comment]
Nice find, I think we can make that change to the 127.0.0.0/8 loopback range for access control in the default config.
